Add an article about VSCode and color palette note (#54)
nikita-rudenko/nikitarudenko.dev: content/blog/mastering-vscode.md

---
date: 2020-11-14T15:13:58Z
tags:
- tooling
- guide
title: Mastering VSCode
excerpt: Sharing my experience with the best code editor

---
VSCode is an ultimate multi-purpose code editor that can fit in any workflow. It has a lot of hidden features and a huge variety of extensions that can take your work experience to a new level. It’s hard to find and keep track of everything, so in this post I will share my experience. The guide is a bit JavaScript/TypeScript centric, however, I believe everyone can pick up something new.

## Shortcuts

I am sure you already know the basic shortcuts for editing, and I want to focus on the most interesting and useful ones. I will use **shortcut names** since the actual keybindings are different depending on your OS. You can open Keyboard Shortcuts in your VSCode instance by clicking the cogwheel in the Activity Bar and look up the combination.

### Essentials

**Show All Commands** - the most powerful shortcut which opens a search through all commands. You can explore and execute any command from here.

**Go to file...** - opens a search through all files and paths in your workspace. Also, shows a list of recently opened files/tabs.

> ### Editing essentials

Reviewer: ```suggestion
### Editing Essentials
```

**Trigger Suggest** - invokes Intellisense. I love using it to discover possible props and destructible values from objects.

**Copy line up/down** - not just line as the name states! Also works for copying selections.

**Go to Bracket** - underrated shortcut for Javascript that helps to untangle parenthesis and brackets by jumping between matching brackets.

**Quick Fix... -** triggers suggestions for refactoring and fixes if the lightbulb is shown next to the gutter

**Format Document** - makes messy code look better.

### **Refactor Inspect**

**Go to Definition** - peek definition and its references.

**Rename Symbol** - rename value and all its occurrences.

**Find** - perform a search in the current file.

**Replace** - find and replace in the current file.

**File: Compare Active File with Saved** - open a diff of the last saved file state and unsaved changes.

> ### **Tab management**

Reviewer: ```suggestion
### **Tab Management**
```

**Switch tabs** - opens a list of currently opened tabs which is great for iterating between files you are working on.

> **Reopen closed tab** - prevents your frustration after closing a tab you didn't want to.

Reviewer: ```suggestion
**Reopen closed tab** - helps you overcome frustration after closing a tab you didn't want to.
```

### **History-based Navigation**

**Go back** - return to the last editing location or tab.

> **Go forward** - undo for **Go back.**

Reviewer: ```suggestion
**Go forward** - undo for **Go back**.
```

**Go to Last Edit Location** - return to the place of your last edit after traveling through imports.

> ### **Quick access**

Reviewer: ```suggestion
### **Quick Access**
```

**View: Toggle Integrated Terminal** - focus, hide, or show the terminal inside VSCode.

**View: Show Search** - jump to the global search across all files in your workspace.

> Additionally, there are [editing hacks](https://code.visualstudio.com/docs/getstarted/tips-and-tricks#_editing-hacks) in the VSCode docs that worth learning.

Reviewer: ```suggestion
Additionally, there are [editing hacks](https://code.visualstudio.com/docs/getstarted/tips-and-tricks#_editing-hacks) in the VSCode docs that are worth learning.
```

![Laptop.](https://res.cloudinary.com/nikita-rudenko/image/upload/v1605476102/nikita-rudenko.dev/laptop.jpg "Photo by Blake Connally on Unsplash")

## Extensions

VSCode utilizes the addon-first approach. Even the basic features are based on extensions (search for "@builtin" in the extensions tab). It means that before starting to work with some new technology, you should consider looking up extensions for it first.

Sharing my list of the most useful and carefully selected extensions I use in my daily workflow:

[**IntelliCode**](https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode) - gives AI boost to Intellisense and tries to make suggestions more useful.

> [**GitLens**](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens) - an essential Swiss army knife for work with Git.

Reviewer: ```suggestion
[**GitLens**](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens) - an essential Swiss army knife for working with Git.
```

[**Live Share**](https://marketplace.visualstudio.com/items?itemName=MS-vsliveshare.vsliveshare) - a killer VSCode extension that takes pair programming to the highest level.

[**Resource Monitor**](https://marketplace.visualstudio.com/items?itemName=mutantdino.resourcemonitor) - a small dashboard with live stats of your machine.

[**Grammarly (unofficial)**](https://marketplace.visualstudio.com/items?itemName=znck.grammarly) - a popular grammar checker useful for Markdown.

[**Test Explorer UI** ](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer)- adds a tree of tests and can be extended with complementary extensions for various testing tools.

**Favorite themes:** [**OneDark Pro**](https://marketplace.visualstudio.com/items?itemName=zhuangtongfa.Material-theme), [**GitHub theme**](https://marketplace.visualstudio.com/items?itemName=GitHub.github-vscode-theme), [**Secunda**](https://marketplace.visualstudio.com/items?itemName=ruj.secunda), [Kaia Theme](https://marketplace.visualstudio.com/items?itemName=ryan0x200.kaia-theme-vscode).

> **Favorite icon themes:** [**Material Icon theme**](https://marketplace.visualstudio.com/items?itemName=PKief.material-icon-theme), [**Awoo**](https://marketplace.visualstudio.com/items?itemName=metalcanine.awoo)**.**

Reviewer: ```suggestion
**Favorite icon themes:** [**Material Icon theme**](https://marketplace.visualstudio.com/items?itemName=PKief.material-icon-theme), [**Awoo**](https://marketplace.visualstudio.com/items?itemName=metalcanine.awoo).
```

#### Front-end extensions:

[**ESLint**](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) - essential static analysis tool that prevents huge amount of potential bugs. I can't imagine my life without ESLint.

> [**Prettier**](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) - essential code formatter that helps keep codebase clean. Can be used in combination with ESLint so formatting errors are shown as lint errors ([guide](https://medium.com/@yaroslavberkut/prettier-eslint-for-react-how-to-kiss-f309789681a6)).

Reviewer: ```suggestion
[**Prettier**](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) - essential code formatter that helps to keep codebase clean. It can be used in combination with ESLint so formatting errors are shown as lint errors ([guide](https://medium.com/@yaroslavberkut/prettier-eslint-for-react-how-to-kiss-f309789681a6)).
```

> [**Quokka**](https://marketplace.visualstudio.com/items?itemName=WallabyJs.quokka-vscode) - a Javascript runner used for creating and debugging algorithms
> 
> [**Debugger for Chrome**](https://marketplace.visualstudio.com/items?itemName=msjsdiag.debugger-for-chrome)  - run debugger right inside VSCode
> 
> [**Document This**](https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis) - a helper for writing JSDoc notations
> 
> [**Jest Runner**](https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner) - easily run and debug Jest tests

Reviewer: Full stops 🔍 
```suggestion
[**Quokka**](https://marketplace.visualstudio.com/items?itemName=WallabyJs.quokka-vscode) - a Javascript runner used for creating and debugging algorithms.

[**Debugger for Chrome**](https://marketplace.visualstudio.com/items?itemName=msjsdiag.debugger-for-chrome)  - run debugger right inside VSCode.

[**Document This**](https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis) - a helper for writing JSDoc notations.

[**Jest Runner**](https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner) - easily run and debug Jest tests.
```
